### ISSUE TITLE
Open API specification document: Contradictory Parent schemas in component section

### ISSUE THREAD

**Issue by mkhan-akhan** [stablekernel/aqueduct#607]
[orm model.zip](https://github.com/stablekernel/aqueduct/files/2705151/orm.model.zip)
The attached two orm classes are referenced in their respective controllers. Aqueduct document command generates the 
[api.zip](https://github.com/stablekernel/aqueduct/files/2705155/api.zip) file. It has two contradictory schemas, one at "components/schemas/LanguageLesson/properties/languageLevel" and the other at "components/schemas/LanguageLevel". The openapi-generator-cli is creating the client class from the incomplete "languageLevel" schema. When this schema is replaced by {"$ref": "#/components/schemas/LanguageLevel"}, the openapi-generator-cli is creating the client class from the complete and correct "LanguageLevel" schema.

I raised this issue in stackoverflow, where it was partially answered. I have posted my findings there as an answer.

It seems like the primary key only schema of a parent class in a child class schema is geared towards fulfilling ORM requirements, however such a definition contradicts the $ref requirement by the Open API Specification 3.0.2, if the full parent schema exists within the components section.

From my understanding so far of Aqueduct (1 weeks worth :), this is not a data issue. I am able to get granular control of data in a parent children relationship (one to many) through join and returningProperties methods.

**Comment by joeconwaystk** [comment]
I'm not quite sure where to start understanding this. The specification I see on StackOverflow shows 2 schema components. One of those schema components has a property named `languageLevel`, but that is a simple object - not part of the schema components. Putting the document into https://editor.swagger.io appears to parse this correctly. What is the expected behavior?

**Comment by mkhan-akhan** [comment]
Sorry, l'll focus on your question regarding expected behaviour. Perhaps we can then proceed from there. The spec file in api.zip that I attached previously is what aqueduct is producing at the moment. The expected behaviour, from my assessment of the client model generated from it, is in this zip file 
[sa_api_expected.zip](https://github.com/stablekernel/aqueduct/files/2706369/sa_api_expected.zip)
The difference is at "components/schemas/LanguageLesson/properties/languageLevel". Once you have looked at this, we can discuss further. Thank you for your patience.

**Comment by mkhan-akhan** [comment]
Regarding your question on whether it is only the Dart generator that declares a new Dart type for properties: the two executions I have done are as follows:
java -jar openapi-generator-cli.jar generate -i sa_api_issue.json -l dart -o openapi_issue -c config.json
and
java -jar openapi-generator-cli.jar generate -i sa_api_issue.json -l csharp -o openapi_csharp -c config.json

The csharp option has the same issue as the dart option.

**Comment by mkhan-akhan** [comment]
I have updated the gist with three files prefixed "title_". I changed the title at components/schemas/LanguageLesson/properties/languageLevel to languageLevelPartial. The generator has created the expected LanguageLevel class, however there is now a new LanguageLevelPartial class which is referenced in the LanguageLesson class.

**Comment by joeconwaystk** [comment]
Got it, the generators just use the title of any object to generate a class, regardless of any other class existing with that name. Hmm. There was a reason that belongs-to relationships were changed not to be a schema reference by default: the typical scenario is that there really only is a simple object with `id` key, and if you are joining on that property, you'd override it. 

I'll have to think on this. What do you think is the right solution?

**Comment by mkhan-akhan** [comment]
Yes, need to think on this too. But just quickly off my head, is it possible to have a flag against the aqueduct document command that instructs it not to use the default belongs-to definition, but to use the schema reference instead.

**Comment by mkhan-akhan** [comment]
As a quick fix, so I can carry on with my app development without having to make manual changes each time a regenerate the api, I have written a function that processes the file generated by aqueduct document and replaces, in the component section, properties of type object with a reference to its corresponding object. See fix_api.dart in the gist https://gist.github.com/mkhan-akhan/b371222b039b906d632e85c20000e966

**Comment by mkhan-akhan** [comment]
> Got it, the generators just use the title of any object to generate a class, regardless of any other class existing with that name. Hmm. There was a reason that belongs-to relationships were changed not to be a schema reference by default: the typical scenario is that there really only is a simple object with `id` key, and if you are joining on that property, you'd override it.
> 
> I'll have to think on this. What do you think is the right solution?

Was the reason behind the above implementation to resolve complex/heavy object 'lifting'. If it is a lazy loading issue, then is a solution based on virtual proxy object possible? Thanks.

**Comment by joeconwaystk** [comment]
No, there is currently a “virtual object” that stands in for undiscovered objects (search “typeref” in package source) and gets resolved once all objects have been documented. The behavior of not using a schema reference object for foreign keys was intentional - all other keys in a nested belongs-to object are ignored by default, so the api document indicates that behavior.
